src/fhirtypes/ElementDefinition.ts

```diff
@@ -2181,10 +2181,15 @@ export class ElementDefinition {
    */
   private cloneChildren(targetElement: ElementDefinition): ElementDefinition[] {
     return targetElement?.children().map(e => {
-      const eClone = e.clone();
+      // When an element is a slice, all constraints should be carried on into the diff
+      // so we do not clear the original
+      const isSlice = e.sliceName != null;
+      const eClone = e.clone(!isSlice);
       eClone.id = eClone.id.replace(targetElement.id, this.id);
       eClone.structDef = this.structDef;
-      eClone.captureOriginal();
+      if (!isSlice) {
+        eClone.captureOriginal();
+      }
```

Comment:
You've implemented this behavior in `cloneChildren`, which is used in places _other than_ unfolding a slice (e.g, unfolding content references uses it too).  While it's probably harmless (as I don't think a content reference can point to a slice), it's behavior that really only applies to a particular use case -- and _could_ be unexpected if we ended up using `cloneChildren` in future code for a different use case.

I think it would be more targeted and less surprising to add a `resetOriginal` or `recaptureOriginal` argument to the `cloneChildren` signature (defaulting to `true`, the previous behavior).  Then in the code that calls this function specifically for the purpose of unfolding a slice, pass in `false` to get the behavior we want there.

Reply:
This comment makes sense, I'll update it.

In response to your top level comment, I can definitely add the logic to `addSlice`, and in face I originally had it there as well. But I couldn't come up with an example where I actually needed that logic in `addSlice`. So I was having a hard time justifying adding this to `addSlice`, which is a relatively significant change to how we handle slices, when I couldn't figure out why exactly we should do that. Although I guess maybe it makes sense just for consistency? Or do you have a case where we can point to an issue in the publisher where we definitely want to have this logic in `addSlice`?

Reply:
> which is a relatively significant change to how we handle slices

Is it any more significant than what we're doing in the unfolding logic now?  I'm just trying to understand if there is something I'm missing.  Or do you just think of it as more significant because it _always_ affects slices, vs unfolding which only affects them when there are child elements?

> Although I guess maybe it makes sense just for consistency?

Yes, I think so.  Unless there is some weirdness that I'm not anticipating.  For example, we obviously would _not_ want the `slicing` to show up in the diff of every _slice_ (even though it is a constraint on the slice root).

> do you have a case where we can point to an issue in the publisher where we definitely want to have this logic in `addSlice`?

I'm not sure that I do.  The problem arises when a slice's invariant/discriminator relies on some constraint introduced in the slice root.  In theory, since FHIRPath has a `elementDefinition()` function that references the ED itself, then I guess it's possible for an author to create an invariant that relies on any arbitrary constraint in the ED (for example the binding).  I'm not sure what real world use case would ever require it, but... if they did, then they'd need all the constraints to be there (I guess).  When it comes down to it, though, I guess it is just for consistency, and because of [this exchange](https://chat.fhir.org/#narrow/stream/179252-IG-creation/topic/How.20to.20slice.20on.20the.20value.20of.20an.20extension/near/257451464) w/ Grahame:

> **Chris Moesel:** Are there any other constraints or ElementDefinition fields that we also should redeclare (i.e. copy) from the base when people create slices of an element?
> **Grahame Grieve:** I think pretty much everything

Reply:
Ok @cmoesel, I've added this in https://github.com/FHIR/sushi/pull/955/commits/70d8784e7f630393ca10ef7337b336fc9d50f0e7, but be aware it does cause a fair bit of regression diffs. I looked through and what I was seeing was all expected.

Reply:
@ngfreiter -- I just noticed another thing when reviewing the code.  Technically, you're only copying the constraints on the element with the sliceName; but if that element has children, those children are technically part of the slice -- so shouldn't they also have copied constraints?

For example:
```
Profile: MyObservation
Parent: Observation
* component MS
* component.code MS
* component ^slicing.discriminator.type = #pattern
* component ^slicing.discriminator.path = "code"
* component ^slicing.rules = #open
* component ^slicing.description = "Slice based on the component.code pattern"
* component contains Foo 0..1
* component[Foo].code ^short = "Foo Code!"
```

When I run this, `Observation.component:Foo` gets the `MS` flag in the differential, but `Observation.component:Foo.code` does not get its `MS` flag in the differential.  So it seems like when we are unfolding children, if the elements to unfold are in a slice, we ought to be preserving/copying constraints there too.  BUT... if there are some unfolded children that didn't get additional constraints in the slice, then we don't want those to show up in the differential at all -- we only want the ones that changed in the slice too.  Which is probably tricky to implement.

Bleh.  This is getting more complicated than I'd like.  I wish the IG Publisher just handled this better.  I am running the regression now to see what kind of churn it makes -- but if you fix the issue I just raised in the previous paragraph, that will surely create even more churn.  Ugh.  Time to go home.

Reply:
We don't need to put those constraints on the unfolded children of the slice do we? Applying `MS` on `Observation.component.code` indicates that all slices of `Observation.component` have `.code` set as `MS`. Wouldn't it be becoming somewhat an abuse of the "differential" to be putting those constraints in children of slices? It feels like an abuse of the differential already to be putting this info in slices when it technically shouldn't be needed. But at least in that case we have a bug we can point to and say, this is why we had to add this stuff to the differential.

Reply:
I guess I don't see how it is different.  Either the slice carries the constraints it inherited or it doesn't.  An element three levels deep in the slice is still part of the slice.  The fact that the current code picks up extensions in a slice is actually just a happy by-product of the fact that extensions happen to have a sliceName -- otherwise they wouldn't have been copied over either.

I agree that the differential shouldn't need to have these things.  But it also seems like if we're going to make slices carry copies of all the constraints, then we should implement it fully (for the whole slice).  But... I do fear what that would mean.

I'll go back to that Zulip thread and try to get some clarification.

Reply:
Yeah I guess this is just me continuing to disagree on the consistency front. To me the original fix (just https://github.com/FHIR/sushi/pull/955/commits/4a3509dff0988566b2275322e460eb8dc5e806e8) is us making an exception to how we do things to satisfy the publisher on something that it seems like the publisher should handle. But to me consistency doesn't justify expanding that exception and creating all this churn.

Reply:
I hear you.  TBH, don't be terribly surprised if you win me over to your side and I ask you to make this work *only* for extensions. ;-)